core/src/main/java/org/apache/iceberg/rest/RESTUtil.java

```diff
@@ -144,7 +163,23 @@ public static Map<String, String> decodeFormData(String formString) {
    */
   public static String encodeString(String toEncode) {
     Preconditions.checkArgument(toEncode != null, "Invalid string to encode: null");
-    return URLEncoder.encode(toEncode, StandardCharsets.UTF_8);
+    return URLEncoder.encode(toEncode, StandardCharsets.UTF_8)
+        // Encode '+' as '%20' to improve compatibility with RFC-3986 which governs encoding for URI
+        // paths and
+        // requires that spaces be encoded as '%20' instead of '+'. URLEncoder#encode encodes
+        // according
+        // 'x-www-form-urlencoded' which encodes spaces as '+'. That's valid for form data and URI
+        // query parameters,
+        // but disagrees with RFC-3986 which considers '+' a reserved character and won't decode to
+        // space.
+        //
+        // Encoding spaces as '%20' is valid both in RFC-3986 and 'x-www-form-urlencoded' and thus
+        // works for path
+        // parameters and query parameters and form data. URLDecoder#decode will decode '%20' to
+        // space.
+        //
+        // Note that the '+' character would be encoded as '%2B' before this replacement.
+        .replace("+", "%20");
```

Comment:
Arguably a better way to increase compatibility with RFC-3986 would be to use a utility that implements RFC-3986 (like Apache Common's [`PercentCodec`](https://github.com/apache/commons-codec/blob/4f0540829d9bd6b3e3e5e2a982f5f929fd42d83c/src/main/java/org/apache/commons/codec/net/PercentCodec.java#L40)).

But I rather we be explicit about the way we diverge from `x-www-form-urlencoded`. This way we can write tests to assert that the round-trip (`RESTUtil#encodeString` and `RESTUtil#decodeString`) continuous to work and we don't break.

Reply:
Also trying to keep the change small and not handling all differences between `x-www-form-urlencoded` and RFC-3986. Unhandled characters are `~` and `*`. See also [JDK-8204530](https://bugs.openjdk.org/browse/JDK-8204530).